### DISCUSSION TITLE
Chiral Scrolling

### DISCUSSION THREAD

**Discussion by poire-z** [koreader/koreader#7810] [Ideas]
![image](https://user-images.githubusercontent.com/24273478/120932445-0c90c200-c6f6-11eb-984b-3ced137c5249.png)

No, joking, we don't need chiral scrolling in KOReader :)

Just mentionning this ChiralScroll Windows app I discovered 2 days ago, and that I'd been wanting for years!:
https://www.reddit.com/r/windows/comments/mxb5mu/i_wrote_a_chiral_scroll_application_for_w10/
https://sourceforge.net/projects/chiralscroll/

Circular/Chiral scrolling had been a feature of Synaptics and Elan touchpads for decades, but since Windows 10 and its own "precision touchpad", touchpad drivers don't (or can't, thanks Microsoft...) provide any other feature any longer... So, it got hard finding a new laptop where I could get it, usually fiddling with drivers, installing old drivers versions.... which I didn't succeed on a new laptop.... So, googled again, and found this, which has been made 1 month ago ! And it works great, even better than some native implementations !

Anyway, I wish to contact the author, but don't want to make a reddit or sourceforge account - but I see he might have an account on GitHub. So, pinging @kered13 - hoping that is you.
(No PM on GitHub ? couldn't think of any other solution than pinging him in some issue or "discussion".)

First, thank you a lot for this !

Second, I noticed some issue on one laptop, where it would not work on some launches. I have to close and restart it a few times for it to work.
Had to install Visual Studio to compile it with your latest commits, hoping they would fix it, but no. So started adding some printf... In case it can help you finding a proper fix:
When it works fine:
```
num_value_caps 17ok
onInit 1 devices
ignored
not stopped, got data
not data
not stopped, got data
maybedata
handlerawinput
numUsages: num 10
numUsages: num 2 status: 1114112
GetButton 66 == 66
GetButton 71 == 66
handlerwainput processTouch
ProcessTouch
Shouldstatscrolling 1 0 1 1
ProcessTouch touchSession shouldstatscrolling
not stopped, got data
```
When it doesn't work;
```
preparsedSize: \\?\HID#DRIVERENUM&Col02#6&a1cfbac&0&0001#{4d1abcde-f16f-11cf-8acb-001111000030} 3572
num_value_caps 17ok
onInit 1 devices
ignored
not stopped, got data
maybedata
handlerawinput
numUsages: num 10
numUsages: num 0 status: 1114112    <=== 0 usage, but STATUS=SUCCESS
handlerwainput processTouch
ProcessTouch
Shouldstatscrolling 1 0 0 1
ProcessTouch touchSession !shouldstatscrolling
not stopped, got data
```
It happens that when inspecting input events, we don't get any "usage" to notice it's a touch... so it never starts scrolling :/
I'm not familiar with Windows API and internals, and couldn't find a proper way to fix it.
I ended up with a small patch that just restarts the app when it notices the "doesn't work" situation :)
Sharing it just in case you feel it might help other users (I see on the reddit post other users might be experimenting it too) : [ChiraScroll_my20210605.patch.txt](https://github.com/koreader/koreader/files/6604497/ChiraScroll_my20210605.patch.txt)

Btw, it would be nice if this app was on GitHub, so it doesn't ever die and people can fork and contribute at keeping it alive !
(We might be only 7 persons in the world really missing it... given the small number of search results of people crying about its loss :/ but I'm sure there are 47 others that have been crying silently ;)

**Comment by Kered13** [comment]
Yes, this is my account. I'm glad you like it. : )

I am aware that there are issues with some laptops. Unfortunately this is the problem with only having one laptop to test on. I have been able to fix some of these issues working with another user on Reddit, but I know they're not all fixed yet. This particular issue is new to me though, so thank you for bringing it to my attention. I don't want to merge in a hack like restarting the application, but I would like to investigate the issue further and see if we can't find a proper solution.

To start with, do you have a patch or diff for where you added those logging statements? It would help me understand the logs you have shared. In the meantime, I was actually planning to add some additional logging to the release version myself to help investigate issues.

Also, if you don't mind making a SourceForge account, it would be helpful if you could file a ticket to track this issue.

PS: It's hosted on SourceForge because I use Mercurial for all of my personal projects (I don't like git in comparison and only use it for contributing to Github projects). Unfortunately SourceForge is pretty much the only free source code hosting site for Mercurial (BitBucket used to provide hosting, but removed all of their support for Mercurial about a year ago).

**Reply by okibcn** [reply]
I missed this feature a lot. It would be nice if you make it open code here on githup.

**Reply by Kered13** [reply]
I use Mercurial for my projects so unfortunately Github is not an option. You can find it hosted on SourceForge [here](https://sourceforge.net/projects/chiralscroll/).

**Comment by poire-z** [comment]
> Also, if you don't mind making a SourceForge account, it would be helpful if you could file a ticket to track this issue.

Well, I did mind for a few minutes, waited a bit, and then stopped minding ;)
https://sourceforge.net/p/chiralscroll/tickets/1/
